qcom: driver init

vpachkov opened the pull request:
Started working on Qualcomm GPU driver...

```
****** create eye Tensor and copy it to the GPU ******
****** initial values ******
values = (1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0)
****** alloc and fill shader ******
****** alloc and fill consts ******
****** fill cmd buff 1 ******
****** fill cmd buff 2 ******
****** submit commands to GPU ******
****** result values ******
values = (2.0, 0.0, 0.0, 0.0, 2.0, 0.0, 0.0, 0.0, 2.0)
```

nimlgen commented:
Cool! Got a bit different output on comma, but something changing in buffers :)

```
****** create eye Tensor and copy it to the GPU ******
****** initial values ******
values = (1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0)
****** alloc and fill shader ******
****** alloc and fill consts ******
****** fill cmd buff 1 ******
****** fill cmd buff 2 ******
****** submit commands to GPU ******
****** result values ******
values = (0.0, 0.0, 0.0, 0.0, 2.0, 0.0, 0.0, 0.0, 2.0)
```

geohot commented:
@vpachkov welcome! you want a bounty lock on this?

geohot commented on extra/qcom_gpu_driver/qcom.py:
```diff
@@ -0,0 +1,174 @@
+    print('****** fill cmd buff 1 ******')
+    # Setting up GPU for the compute shader mode...
+    buff1 = CMDBuffer()
+    buff1.push(opcode=0x26) # CP_WAIT_FOR_IDLE
```
Ideally there's ways to autogen a header file with all these constants instead of comments like this.

vpachkov commented:
@geohot Thank you! And sure, I would like to continue with this one.
Based on my current understanding of tinygrad and Adreno, I would expect to finish in 2-3 weeks.

- [ ] Make a header file to represent commands more clear (~ 1 day)
- ioctl and args are presented in msm_kgsl.h
- typ7 commands - adreno_pm4.xml
- registers - a6xx.xml
- [ ] Compiler (~ 1 day)
Dump calls to the shader compiler functions. Found out that OpenCL utilizes libllvm-qcom.so. So then just repeat it.
- [ ] HWPM4Queue + HWCopyQueue (~ 6 days)
- From looking at ops_amd.py, there are 2 types of hardware queues in tinygrad: command queue and copy queue. Thus, create something similar for Qualcomm. (~3 days)
- Implement signals (~ 3 days)
- [ ] Allocator (~ 1 day)
- Almost everything would stay unchanged. Just add function for freeing memory.
- [ ] Program (~ 3 days)
- [ ] Testing (~ 3 days)

geohot commented:
Locked.

You shoudn't need copy queue for Qualcomm, it's unified memory. Just compute is fine. Even without unified, copy queue is really only important for multGPU.

Good call with the shader compiler, I didn't realize this was a unique challenge of Qualcomm. If it's annoying, you are also welcome to just use OpenCL and export the binary.

geohot commented:
We have a comma 3 available to you over SSH when you are ready to test on it.

nimlgen commented on tinygrad/runtime/ops_qcom.py:
```diff
@@ -0,0 +1,148 @@
+    alloc = kgsl.struct_kgsl_gpuobj_alloc(size=size, flags=flags)
+    self._ioctl(kgsl.IOCTL_KGSL_GPUOBJ_ALLOC, alloc)
+    info = kgsl.struct_kgsl_gpuobj_info(id=alloc.id)
+    self._ioctl(kgsl.IOCTL_KGSL_GPUOBJ_INFO, info)
```
You can think of switching this to the `info = kgsl.gpuobj_info(id=alloc.id)` style (like amd/nv) with some preprocessing of packets.

nimlgen commented:
@vpachkov, you're welcome to join our Monday meetings (9:40 PT) on Discord to discuss this pr and hcq-related things.

geohot commented:
Is this still active @vpachkov?

geohot commented:
How's progress on this? We have a new ETA of completion?

vpachkov commented:
> How's progress on this? We have a new ETA of completion?
> 
> Sorry to be pushy on this, but we need it for comma. Will up the bounty to $1,000 if it's merged by next Wednesday

It's really hard to extract info that is needed to run a shader from the row binary... Such as HALFREGFOOTPRINT, FULLREGFOOTPRINT and other stuff. Right now I use a modified parser from freedreno that is compiled from tinygrad into a .so (like it's done for disasm), but it is a mess

vpachkov commented:
this can now run some tests!

```
~/tinygrad $ QCOM=1 python test/test_ops.py TestOps.test_tiny_add
test_tiny_add (__main__.TestOps.test_tiny_add) ...
testing                                   [3, 3]   np/tinygrad fp: 0.02 / 685.06 ms  bp: nan / nan ms ok

----------------------------------------------------------------------
Ran 1 test in 0.712s

OK
~/tinygrad $ QCOM=1 python test/test_ops.py TestOps.test_tiny_mul
test_tiny_mul (__main__.TestOps.test_tiny_mul) ...
testing                                 [64, 64]   np/tinygrad fp: 0.02 / 766.84 ms  bp: nan / nan ms ok

----------------------------------------------------------------------
Ran 1 test in 0.794s

OK
```

geohot commented:
For all these things, a good strategy is to just use the sniffer and copy exactly what it's doing

vpachkov commented:
this can now run kernels with constants and custom arguments layouts!
```
import tinygrad as tg
x = tg.Tensor([1.0, 2.0, 3.0, 4.0, 5.0])
y = x.sin()
print(y.numpy())
```
```
~/tinygrad $ QCOM=1 python tinygrad/runtime/ops_qcom.py
[ 0.84147096  0.9092974   0.14112002 -0.7568025  -0.9589243 ]
```

vpachkov commented:
> For all these things, a good strategy is to just use the sniffer and copy exactly what it's doing

Sure, but this step was done at least a week ago. From that time, I'm working on a generic solution that can run any kernel it was passed to run. And does not hang the gpu :)

vpachkov commented:
Running gpt2 on qcom!

```
vpachkov@tiny22:~/tinygrad$ QCOM=1 python examples/gpt2.py
using QCOM backend
using gpt2-medium
ram used:  1.42 GB, lm_head.weight                                    : 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████| 293/293 [00:10<00:00, 28.47it/s]
loaded weights in 10297.32 ms, 1.63 GB loaded at 0.16 GB/s
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 100/100 [01:57<00:00,  0.85it/s]
Generating text...
What is the answer to life, the universe, and everything?

(V.S., October 19, 1994.)

Professor of English,

I am writing to inform you that I am a father of a child born in September of 1994. I mentioned the name of these children to your staff director and he said: "I have a wife who is a widower and the son of a widower. We usually take a baby to a hospital when there is something urgent other than a flat (two or three pints of milk being one
```

nimlgen commented on tinygrad/runtime/ops_qcom.py:
```diff
@@ -0,0 +1,269 @@
+    submit_req.context_id = device.ctx
+
+    device._ioctl(0x4A, submit_req)
+    self.q = []
```
remove this

vpachkov replied:
done

nimlgen commented on tinygrad/runtime/ops_qcom.py:
```diff
@@ -0,0 +1,269 @@
+    self.cmd(adreno.CP_RUN_OPENCL, 0)
+    self.cmd(adreno.CP_WAIT_FOR_IDLE)
+
+class QcomCopyQueue(HWCopyQueue, QcomComputeQueue):
```
copy queue will be optional in master soon

vpachkov replied:
Nice! Done

nimlgen commented on tinygrad/runtime/ops_qcom.py:
```diff
@@ -0,0 +1,291 @@
+        0xccc0cf, 0x2fc,
+        global_size[0], global_size[1], global_size[2], # original global sizes
+    )
+    self.cmd(adreno.CP_SET_MARKER, adreno.RM6_COMPUTE)
```
can we move this into device's `__init__`?

vpachkov replied:
done

vpachkov commented:
gpt2 on QCOM + HCQ is a little faster than GPU: `ran model in 305.47 ms` vs `ran model in 315.65 ms`

nimlgen commented on tinygrad/runtime/ops_qcom.py:
```diff
@@ -0,0 +1,291 @@
+      return
+
+    cmdbytes = array.array('I', self.q)
+    alloc = device._gpu_alloc(len(cmdbytes) * 4, 0xC0A00, map_to_cpu=True)
```
never freed. consider a ring buffer here?

vpachkov replied:
done

nimlgen commented on tinygrad/runtime/ops_qcom.py:
```diff
@@ -0,0 +1,310 @@
+  def __init__(self, device: QcomDevice, name: str, lib: bytes):
+    self.device, self.name, self.lib = device, name, lib
+
+    self.private_gpu = self.device._gpu_alloc(0x101, 0xC0F00)
```
private mem should be shared for all programs, no?

vpachkov replied:
yeah, just noticed it too. done.

nimlgen commented on tinygrad/runtime/ops_qcom.py:
```diff
@@ -0,0 +1,310 @@
+    ctypes.memmove(self.lib_gpu.va_addr, mv_address(memoryview(image)), self.image_size)
+
+    super().__init__(self.device, self.name, kernargs_alloc_size=0x1000, kernargs_args_offset=self.buffs_info[0] if len(self.buffs_info) else 0)
+
```
`__del__`

vpachkov replied:
done

geohot commented on tinygrad/runtime/graph/hcq.py:
```diff
@@ -11,6 +11,7 @@ class HCQGraph(MultiGraphRunner):
   def __init__(self, jit_cache: List[ExecItem], input_rawbuffers: List[Buffer], var_vals: Dict[Variable, int]):
     super().__init__(jit_cache, input_rawbuffers, var_vals)
     self.devices = list(set(cast(HCQCompiled, d) for ji in jit_cache for d in [Device[cast(Buffer, x).device] for x in ji.bufs]))
+    self.bufs_stride, self.vars_stride = (2, 4) if self.devices[0].dname.startswith('QCOM') else (1, 1)
```
What is this and why does qcom need it?

vpachkov replied:
This is a hack to support hcq. amd/nv populates its bufs and vars sequentially 0x140, 0x148, ....
qcom does it differently, 0x140, 0x150 and so on. thus, aligning them by 16 bytes.
see hexdump of bufs in qcom constants:
```
00000140: 00 B0 BA FF 07 00 00 00  00 00 00 00 00 00 00 00  ................
00000150: 00 00 BD FF 07 00 00 00  00 00 00 00 00 00 00 00  ................
```
Of course, it's not the best decision to set these strides here, but hcq api currently does not support exporting this functionality to the device implementation. I think I can come up with something better in a separate PR after this is merged. Or do it just here, if this is too "hacky".

nimlgen replied:
master have an abstraction for this now

vpachkov commented:
Current status:

After excluding tests that fails even with `GPU=1`, running `QCOM=1 python test/test_ops.py` consistently gives the following result:
```
Ran 292 tests in 309.560s

FAILED (errors=11, skipped=20)
```
looks like these tests are failed because of this timeout in the kernel:
```
/*
 * Wait for the last global timestamp to pass before continuing.
 * The maxumum wait time is 30s, some large IB's can take longer
 * than 10s and if hang happens then the time for the context's
 * commands to retire will be greater than 10s. 30s should be sufficient
 * time to wait for the commands even if a hang happens.
 */
ret = adreno_drawctxt_wait_rb(adreno_dev, &drawctxt->base,
	drawctxt->internal_timestamp, 30 * 1000);
```
**all of these failed tests are ok** when running them separately:
```
vpachkov@tiny22:~/tinygrad$ for test in test_grouped_conv_transpose2d test_maxpool2d_dilation test_maxpool2d_smaller_stride test_output_padded_conv_transpose2d test_padded_conv2d_bs1 test_padded_conv_transpose2d test_simple_conv_transpose2d test_simple_conv_transpose3d; do QCOM=1 python test/test_ops.py TestOps.$test; done
test_grouped_conv_transpose2d (__main__.TestOps.test_grouped_conv_transpose2d) ...
testing             [(2, 4, 9, 9), (4, 4, 3, 3)]   torch/tinygrad fp: 8.69 / 1413.26 ms  bp: 12.55 / 762.55 ms ok

----------------------------------------------------------------------
Ran 1 test in 2.381s

OK
test_maxpool2d_dilation (__main__.TestOps.test_maxpool2d_dilation) ...
testing                         [(8, 2, 17, 14)]   torch/tinygrad fp: 3.37 / 1231.40 ms  bp: 3.23 / 1251.07 ms
testing                         [(8, 2, 17, 14)]   torch/tinygrad fp: 1.21 / 89.27 ms  bp: 1.82 / 1076.83 ms
testing                         [(8, 2, 17, 14)]   torch/tinygrad fp: 1.25 / 94.33 ms  bp: 1.82 / 1342.58 ms
testing                         [(8, 2, 17, 14)]   torch/tinygrad fp: 1.24 / 83.72 ms  bp: 1.76 / 509.04 ms ok

----------------------------------------------------------------------
Ran 1 test in 5.766s

OK
test_maxpool2d_smaller_stride (__main__.TestOps.test_maxpool2d_smaller_stride) ...
testing                         [(8, 2, 17, 14)]   torch/tinygrad fp: 2.75 / 1233.43 ms  bp: 3.31 / 1368.88 ms
testing                         [(8, 2, 17, 14)]   torch/tinygrad fp: 1.28 / 88.96 ms  bp: 1.87 / 1495.73 ms
testing                         [(8, 2, 17, 14)]   torch/tinygrad fp: 1.21 / 92.55 ms  bp: 1.88 / 1478.34 ms
testing                         [(8, 2, 17, 14)]   torch/tinygrad fp: 2.16 / 93.97 ms  bp: 1.85 / 1459.01 ms ok

----------------------------------------------------------------------
Ran 1 test in 7.407s

OK
test_output_padded_conv_transpose2d (__main__.TestOps.test_output_padded_conv_transpose2d) ...
testing       [(2, 4, 6, 5), (4, 4, 3, 3), (4,)]   torch/tinygrad fp: 12.18 / 1661.09 ms  bp: 15.72 / 953.29 ms
testing       [(2, 4, 6, 5), (4, 4, 3, 3), (4,)]   torch/tinygrad fp: 1.84 / 528.07 ms  bp: 20.81 / 942.64 ms ok

----------------------------------------------------------------------
Ran 1 test in 4.709s

OK
test_padded_conv2d_bs1 (__main__.TestOps.test_padded_conv2d_bs1) ...
testing           [(1, 3, 11, 28), (4, 3, 3, 3)]   torch/tinygrad fp: 3.02 / 1412.44 ms  bp: 5.49 / 1174.67 ms ok

----------------------------------------------------------------------
Ran 1 test in 2.706s

OK
test_padded_conv_transpose2d (__main__.TestOps.test_padded_conv_transpose2d) ...
testing             [(2, 4, 9, 9), (4, 4, 3, 3)]   torch/tinygrad fp: 15.08 / 1315.47 ms  bp: 16.28 / 665.62 ms
testing             [(2, 4, 9, 9), (4, 4, 3, 3)]   torch/tinygrad fp: 1.96 / 171.05 ms  bp: 13.87 / 646.30 ms
testing             [(2, 4, 9, 9), (4, 4, 3, 3)]   torch/tinygrad fp: 1.94 / 138.12 ms  bp: 14.75 / 535.41 ms
testing             [(2, 4, 9, 9), (4, 4, 3, 3)]   torch/tinygrad fp: 1.64 / 194.90 ms  bp: 13.81 / 1091.69 ms
testing             [(2, 4, 9, 9), (4, 4, 3, 3)]   torch/tinygrad fp: 1.70 / 183.91 ms  bp: 13.80 / 734.93 ms ok

----------------------------------------------------------------------
Ran 1 test in 6.181s

OK
test_simple_conv_transpose2d (__main__.TestOps.test_simple_conv_transpose2d) ...
testing             [(2, 4, 9, 9), (4, 4, 3, 3)]   torch/tinygrad fp: 12.04 / 1330.76 ms  bp: 14.73 / 733.99 ms ok

----------------------------------------------------------------------
Ran 1 test in 2.266s

OK
test_simple_conv_transpose3d (__main__.TestOps.test_simple_conv_transpose3d) ...
testing       [(2, 4, 9, 9, 9), (4, 4, 3, 3, 3)]   torch/tinygrad fp: 7.70 / 1482.01 ms  bp: 16.17 / 1633.10 ms ok

----------------------------------------------------------------------
Ran 1 test in 3.349s

OK
```

Considering the urgency of this PR and the fact that test are ok separately, I open it for review. While trying to fix failed tests when they run all together.

nimlgen commented on test/test_ops.py:
```diff
@@ -344,6 +344,7 @@ def test_lerp(self):
     helper_test_op([(45,35), (45,35), (45,35)], lambda x,y,z: x.lerp(y,z))
     helper_test_op(None, lambda x,y,z: x.lerp(y,z), vals=[[1.,2.,3.], [4.,5.,6.], 0.5])
 
+  @unittest.skipIf(Device.DEFAULT in {"GPU", "QCOM"}, "OpenCL fails to compile this (both on GPU/QCOM backends)")
```
do not skip them on GPU

vpachkov replied:
ok. done.

geohot commented:
How are Images coming along?

vpachkov commented:
> How are Images coming along?

still trying to fix the last test in test_image_dtype, but the rest are working:
```
vpachkov@tiny22:~/tinygrad$ for test in \
>     test_image_and_back_wrong_shape \
>     test_shrink_load_float \
>     test_mul_stays_image \
>     test_sum \
>     test_shrink_max \
>     test_shrink_to_float \
>     test_lru_alloc \
>     test_no_lru_alloc \
>     test_no_lru_alloc_dtype \
>     test_lil_model; do
>     QCOM=1 python test/test_image_dtype.py TestImageDType.$test
> done

.
----------------------------------------------------------------------
Ran 1 test in 1.206s

OK
.
----------------------------------------------------------------------
Ran 1 test in 1.279s

OK
.
----------------------------------------------------------------------
Ran 1 test in 1.175s

OK
.
----------------------------------------------------------------------
Ran 1 test in 1.204s

OK
.
----------------------------------------------------------------------
Ran 1 test in 1.299s

OK
.
----------------------------------------------------------------------
Ran 1 test in 1.296s

OK
.
----------------------------------------------------------------------
Ran 1 test in 1.185s

OK
.
----------------------------------------------------------------------
Ran 1 test in 1.257s

OK
.
----------------------------------------------------------------------
Ran 1 test in 1.267s

OK
[0.0, 0.0, 0.0, 0.0]
[0.0]
[2.0]
[0.5, 0.5]
[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
.
----------------------------------------------------------------------
Ran 1 test in 1.856s

OK
```

vpachkov commented:
published my code sketches for images. added comments and an example that works and another that doesn't. 
also added texture dumping stuff to the sniffer

geohot commented on tinygrad/runtime/ops_qcom.py:
```diff
@@ -0,0 +1,422 @@
+
+    self.cmd(adreno.CP_WAIT_FOR_IDLE)
+    self.reg(adreno.REG_A6XX_HLSQ_CS_NDRANGE_0,
+             (3 << adreno.A6XX_HLSQ_CS_NDRANGE_0_KERNELDIM__SHIFT) | ((local_size[0] - 1) << adreno.A6XX_HLSQ_CS_NDRANGE_0_LOCALSIZEX__SHIFT)
```
@nimlgen Can we get wrapper functions for these? This is hard to read

geohot commented:
@vpachkov getting this merged is now @nimlgen's top priority, and when we do the full bounty is still yours. We need this for comma ASAP.
